design/design-generic-references.md

```diff
@@ -0,0 +1,477 @@
+looks only for the value in `LocationS3` instance named
+by `spec.forProvider.sourceLocationArnRef` field.
+
+Currently, the issue can be solved by defining the following composition:
+
+* Declare under managed resources the needed source object (one of `Location*`)
+  types with [observe only management policy]
+* Patch composite object with the value from the source object
+* Patch Task's `spec.forProvider.sourceLocationArn` field with the composite
+  value
+
+The proposed solution has the following drawbacks:
+
+* For each claim, additional observe only resource gets created, just to be able
+  to work around the limitations of the current reference mechanism on certain
+  fields. In a system with a high number of claims, that would put additional
+  pressure on k8s API and etcd.
+* Composite object needs to expose the patched value, although fully unneeded
+  from API design perspective
+
+__**NOTE**__: [a prior design exists on this topic](https://github.com/crossplane/crossplane/pull/2385).
+
+## Goals
+
+The generic cross-reference should be able to:
+
+* Refer a value on an arbitrary path within an object of arbitrary kind
+* Use the existing [Kubernetes API object reference] convention as much as
+  possible
+* Be implementable without or requiring only light changes in RBAC
+
+In cases when the generic cross-references are unavailable for a given managed
+resource field, our compositions should not require declaring observe-only
+resource, for sole purpose of referring a value.
+
+## Proposal
+
+Until now, providers were supporting only reference fields of
+type [`Reference`](https://github.com/pedjak/crossplane-runtime/blob/master/apis/common/v1/resource.go#L116-L123):
+
+```go
+type Reference struct {
+// Name of the referenced object.
+Name string `json:"name"`
+
+// Policies for referencing.
+// +optional
+Policy *Policy `json:"policy,omitempty"`
+}
+```
+
+### GenericReference Type
+
+We would like to introduce `GenericReference` field type:
+
+```go
+type GenericReference struct {
+  Reference
+  
+  // ApiVersion of the referenced object.
+  ApiVersion string `json:"apiVersion"`
+  
+  // Kind of the referenced object.
+  Kind string `json:"kind"`
+  
+  // FieldPath of the value within the referenced object.
+  FieldPath string `json:"fieldPath"`
+}
+```
+
+The syntax of the `fieldPath` field is the one used in Composition, in line with
+the [Kubernetes API Conventions on field selection].
+
+Reference fields of the above type describe completely the location of 
+the value that is about to be set on the counterpart field. In case of 
+`datasync.aws.upbound.io` `Task`, an instance might look like:
+
+```yaml
+apiVersion: datasync.aws.upbound.io/v1beta1
+kind: Task
+metadata:
+  name: datasync-task-example
+spec:
+  forProvider:
+    .
+    .
+    sourceLocationArnRef:
```

Comment:
Does this mean that provider developers still need to explicitly identify and declare what fields can be set by a reference at build time? For example here the developer would need to identify that `sourceLocationArn` could reference another type, and thus make sure `sourceLocationArnRef` existed.

Reply:
> Does this mean that provider developers still need to explicitly identify and declare what fields can be set by a reference at build time? For example here the developer would need to identify that sourceLocationArn could reference another type, and thus make sure sourceLocationArnRef existed.

My understanding so far (and chatting with @muvaf) is that we have automated that with Upjet, but I could be ofcourse wrong here.

Reply:
> My understanding so far (and chatting with @muvaf) is that we have automated that with Upjet, but I could be ofcourse wrong here.

I thought we'd automated _generating the code_ for references, but not automated the process of _identifying which fields can be set using a reference_. It's possible I'm wrong too though - CC @ulucinar and @muvaf.

Either way though, I suspect only being able to use references for certain fields will be too limiting. If a field can reference an arbitrary type, how would we know at development which fields should be able to be set by a reference and which can't? Do we think that we'll be able to predict all the fields users might want to derive using a reference at provider development time?

Reply:
> Either way though, I suspect only being able to use references for certain fields will be too limiting. If a field can reference an arbitrary type, how would we know at development which fields should be able to be set by a reference and which can't? Do we think that we'll be able to predict all the fields users might want to derive using a reference at provider development time?

My understanding is that providers generated with the help of Upjet have mapped all fields that can reference external values, correct @ulucinar? These fields are all visible in MR schemas, hence easy discoverable for consumers. Other, non-upjet based providers needs to references in their MR schema in some way as well.

@haarchri, how is your experience? Did you stumble across some fields that should be referable, but they are not according to MR schema?

Reply:
Alper can correct me if I am wrong, but AFAIK, upjet infers some of those references by parsing their documentation and the examples there but not all. For unidentified ones, it is still a manual process to configure them.

Reply:
Yes, correct:
> My understanding is that providers generated with the help of Upjet have mapped all fields that can reference external values, correct @ulucinar?

We attempt to deduce the reference parameters from the examples provided in the Terraform docs and this does *not* result in the generation of all reference parameters. For instance, none of the provided example Terraform configurations could be utilizing a possible reference. We do not attempt to parse the argument documentation for this purpose, only the provided HCL configurations are used.

Thus, we have some automation for determining the reference parameters themselves on top of the reference implementation generator in angryjet. But, it's not 100% reliable, optional (an upjet-based provider's maintainer can opt out not to generate the reference parameters) and to the best of my knowledge, is only applicable for upjet-based providers.

@negz, are you suggesting a referencing scheme where, at runtime, we can declare a reference whose source & target are arbitrary (no a priori knowledge about who can reference what), so that at build time, we don't need to know anything about the references?

My understanding is that the proposed approach here makes the reference source (where the data will copied from) dynamic (GVK+name+fieldpath) but the reference target (where the data will be copied into) is still static, i.e., you need to know the target at build time.